networkx/algorithms/traversal/depth_first_search.py

```diff
@@ -78,12 +85,16 @@ def dfs_edges(G, source=None, depth_limit=None):
     if depth_limit is None:
         depth_limit = len(G)
 
+    get_children = G.neighbors
+    if sort_children is not None and callable(sort_children):
+        get_children = lambda node: iter(sort_children(G.neighbors(node)))
```

Comment:
FWIW I think this makes sense. From the other discussion in the PR, it seems like there's a bit of confusion on whether `get_children` should be an iterable or an iterator. IIUC, the confusion seems to arise from the fact that iterators (e.g. generators) *are* iterables (but not all iterables are iterators!).

So - if `sort_children` here were e.g. a generator, then this code would generate a function `get_children` that returned a generator (i.e. iterable) as desired.

Reply:
if the `sort_children` function returns an iterable(e.g. list) then when we call the `get_children` later, then `sort_children(G.neighbors(node)` will first do all the computation for sorting and return an iterable, and then it will be converted into an iterator by `iter()` function. But, when the `sort_children` returns an iterator the computation will be done here `for child in children:`, when the `children` is used. So, that's why I think it would be slightly better if `sort_children` returns an iterator, as compared to when it will return an iterable.

Reply:
Looking at the `sort_neighbors` inputs within the BFS module, and the way the code here works, I think:
-  sort_children should be called sort_neighbors here even though other parts of the code call it child and children. Those are internal names (I hope) so don't impact outside the function, and in the broader literature child and parent are not consistently one direction or the other in a directed graph. So, neighbor matches the BFS code within NetworkX and avoids the confusion of the words child and parent.
- sort_neighbors (counter to the some doc_strings in BFS) is given an iterator as input -- not an iterable. The docs are pretty clear that it should return an iterator -- but its use allows an iterable. It seems like we should change the docs to match what the code enforces/assumes.
- For both the BFS and DFS it seems good to have the docs say that the function takes an iterator of nodes as input. That is what we do when we apply the function to `G.neighbors` or `G.predecessors`. 
- For both the BFS and DFS it seems good to have the docs say that the function should return an iterable, since that is what we require. (we could ask for it to return an iterator and remove all our calls to `iter(sort_...` but that would be backward incompatible with only very marginal speedup of the code. Memory use is hard to know because the`sort_neighbors` function could easily construct a list from it's input, sort that list and then return an iterator over the list. So the whole "iterator" memory advantage is likely lost whenever sorting is involved.)

The important parts of this discussion (in my view) are:
1) the input to the function is an iterator (can only be passed through once after which it is exhausted).
2) we take `iter(...)` of the output so it can be an iterable (like a list) even though it allows memory use reduction if the function is "thin" and returns an iterator.
3) the docs should reflect items 1) and 2), and I believe they currently don't.